clang/test/SemaSYCL/zero-length-arrays.cpp

```diff
@@ -0,0 +1,162 @@
+      // expected-note@+1 {{field of illegal pointer type 'ZEROARR *' (aka 'float (*)[0]') declared here}}
+      (void)Ptr; // expected-error {{zero-length arrays are not permitted in SYCL device code}}
+    });
+  });
+  // expected-note@#KernelSingleTaskKernelFuncCall {{called by 'kernel_single_task<Simple1, (lambda at}}
+  // expected-error@#KernelSingleTaskKernelFuncCall {{zero-length arrays are not permitted in SYCL device code}}
+  // expected-note@#KernelSingleTask {{in instantiation of function template specialization}}
+  q.submit([&](handler &h) {
+    // expected-note@+2 {{in instantiation of function template specialization}}
+    // expected-note@+1 {{field of illegal type 'int[0]' declared here}}
+    h.single_task<class Simple1>([Arr]{ // expected-error {{zero-length arrays are not permitted in SYCL device code}}
+    });
+  });
+  WrapperOfWrapper St;
+  // expected-note@#KernelSingleTaskKernelFuncCall {{called by 'kernel_single_task<SimpleStruct, (lambda at}}
+  // expected-error@#KernelSingleTaskKernelFuncCall 2{{zero-length arrays are not permitted in SYCL device code}}
+  // expected-note@#KernelSingleTask {{in instantiation of function template specialization}}
+  q.submit([&](handler &h) {
+    // expected-note@+1 {{in instantiation of function template specialization}}
+    h.single_task<class SimpleStruct>([=]{
+      // expected-note@+1 {{within field of type 'WrapperOfWrapper' declared here}}
+      (void)St.F.BadArray; // expected-error 4{{zero-length arrays are not permitted in SYCL device code}}
+    });
+  });
+  // expected-note@#KernelSingleTaskKernelFuncCall {{called by 'kernel_single_task<SimpleStruct1, (lambda at}}
+  // expected-error@#KernelSingleTaskKernelFuncCall 2{{zero-length arrays are not permitted in SYCL device code}}
+  // expected-note@#KernelSingleTask {{in instantiation of function template specialization}}
+  q.submit([&](handler &h) {
+    // expected-note@+2 {{in instantiation of function template specialization}}
+    // expected-note@+1 {{within field of type 'WrapperOfWrapper' declared here}}
+    h.single_task<class SimpleStruct1>([St]{ // expected-error 2{{zero-length arrays are not permitted in SYCL device code}}
+    });
+  });
+
+  Templated<1, int> OK;
+  Templated<1 - 1, double> Weirdo;
+  Templated<0, float> Zero;
+  // expected-note@#KernelSingleTaskKernelFuncCall {{called by 'kernel_single_task<UseTemplated, (lambda at}}
+  // expected-error@#KernelSingleTaskKernelFuncCall 4{{zero-length arrays are not permitted in SYCL device code}}
+  // expected-note@#KernelSingleTask {{in instantiation of function template specialization}}
+  q.submit([&](handler &h) {
+  // expected-note@+1 {{in instantiation of function template specialization}}
+    h.single_task<class UseTemplated>([=]{
+      (void)OK; // No errors expected
+      (void)Zero; // expected-error 2{{zero-length arrays are not permitted in SYCL device code}}
```

Comment:
Same here. Why 2 diagnostics for the same thing? Is the additional diagnostic for instantiated template?

Reply:
Because there are two zero-length arrays inside this object. An error is generated for each of them (notes are not emitted after first error for the same object).